### ISSUE TITLE
ims_registrar_pcscf: support for handling contact * value

### ISSUE THREAD

**Issue by deepak-git-account** [kamailio/kamailio#3848]
### Description

     If Deregistration contains * in a contact field as described in RFC-3261,
     then the Kamailio is not cleaning the in-memory and database record. And the next iteration is not working correctly. 
     Use of the "*" Contact header field value allows a registering UA
      to remove all bindings associated with an address-of-record
      without knowing their precise values.

    Step:1 REGISTER USER UE-A
    Step:2: Wait for the certain duration
    Step:3: De-REGISTER USER UE-A (Contact header is using *)
    Step:4: Repeat from the step-1
    After step:3, clean-up is not happening. Record can be seen in SQL.
   Note: If de-register contains specific CONTACT header, then cleanup is happening.
#### Debugging Data

```
(paste your debugging data here)
```

#### Log Messages

```
May 13 19:47:22 focal pcscf/pcscf[35816]: DEBUG: <script>: Contact headervalue: *
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: *** cfgtrace:dbg_cfg_trace(): request_route=[REGISTER] c=[/usr/local/etc/kamailio/pcscf/route/register.cfg] l=25 a=63 n=assign
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: *** cfgtrace:dbg_cfg_trace(): request_route=[REGISTER] c=[/usr/local/etc/kamailio/pcscf/route/register.cfg] l=26 a=26 n=xlog
May 13 19:47:22 focal pcscf/pcscf[35816]: DEBUG: <script>: Full URI from sip:49404965200004@open-ims.test: sip:49404965200004@open-ims.test
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: *** cfgtrace:dbg_cfg_trace(): request_route=[REGISTER] c=[/usr/local/etc/kamailio/pcscf/route/register.cfg] l=33 a=63 n=assign
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: *** cfgtrace:dbg_cfg_trace(): request_route=[REGISTER] c=[/usr/local/etc/kamailio/pcscf/route/register.cfg] l=34 a=63 n=assign
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: *** cfgtrace:dbg_cfg_trace(): request_route=[REGISTER] c=[/usr/local/etc/kamailio/pcscf/route/register.cfg] l=36 a=63 n=assign
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: *** cfgtrace:dbg_cfg_trace(): request_route=[REGISTER] c=[/usr/local/etc/kamailio/pcscf/route/register.cfg] l=37 a=26 n=xlog
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: *** cfgtrace:dbg_cfg_trace(): request_route=[REGISTER] c=[/usr/local/etc/kamailio/pcscf/route/register.cfg] l=38 a=63 n=assign
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: *** cfgtrace:dbg_cfg_trace(): request_route=[REGISTER] c=[/usr/local/etc/kamailio/pcscf/route/register.cfg] l=39 a=26 n=xlog
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: *** cfgtrace:dbg_cfg_trace(): request_route=[REGISTER] c=[/usr/local/etc/kamailio/pcscf/route/register.cfg] l=46 a=16 n=if
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: *** cfgtrace:dbg_cfg_trace(): request_route=[REGISTER] c=[/usr/local/etc/kamailio/pcscf/route/register.cfg] l=43 a=26 n=xlog
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: *** cfgtrace:dbg_cfg_trace(): request_route=[REGISTER] c=[/usr/local/etc/kamailio/pcscf/route/register.cfg] l=54 a=16 n=if
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: *** cfgtrace:dbg_cfg_trace(): request_route=[REGISTER] c=[/usr/local/etc/kamailio/pcscf/route/register.cfg] l=46 a=25 n=is_present_hf
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: *** cfgtrace:dbg_cfg_trace(): request_route=[REGISTER] c=[/usr/local/etc/kamailio/pcscf/route/register.cfg] l=47 a=26 n=xlog
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: *** cfgtrace:dbg_cfg_trace(): request_route=[REGISTER] c=[/usr/local/etc/kamailio/pcscf/route/register.cfg] l=50 a=25 n=append_hf
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: *** cfgtrace:dbg_cfg_trace(): request_route=[REGISTER] c=[/usr/local/etc/kamailio/pcscf/route/register.cfg] l=54 a=26 n=xlog
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: *** cfgtrace:dbg_cfg_trace(): request_route=[REGISTER] c=[/usr/local/etc/kamailio/pcscf/route/register.cfg] l=55 a=25 n=pcscf_save_pending
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: ims_registrar_pcscf [save.c:318]: save_pending(): No contact headers
May 13 19:47:22 focal pcscf/pcscf[35816]: DEBUG: ims_registrar_pcscf [save.c:492]: save_pending(): Error saving pending contact
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: *** cfgtrace:dbg_cfg_trace(): request_route=[REGISTER] c=[/usr/local/etc/kamailio/pcscf/route/register.cfg] l=67 a=16 n=if
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: *** cfgtrace:dbg_cfg_trace(): request_route=[REGISTER] c=[/usr/local/etc/kamailio/pcscf/route/register.cfg] l=65 a=63 n=assign
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: *** cfgtrace:dbg_cfg_trace(): request_route=[] c=[] l=0 a=12 n=seturi
May 13 19:47:22 focal pcscf/pcscf[35816]: ERROR: *** cfgtrace:dbg_cfg_trace(): request_route=[REGISTER] c=[/usr/local/etc/kamailio/pcscf/route/register.cfg] l=71 a=26 n=xlog


```

#### SIP Traffic

```
(paste your sip traffic here)
```

### Additional Information

  * **Kamailio Version** - output of `kamailio -v`

```
kamailio -v
version: kamailio 5.7.5 (x86_64/linux) 58499a-dirty
flags: USE_TCP, USE_TLS, USE_SCTP, TLS_HOOKS, USE_RAW_SOCKS, DISABLE_NAGLE, USE_MCAST, DNS_IP_HACK, SHM_MMAP, PKG_MALLOC, MEM_JOIN_FREE, Q_MALLOC, F_MALLOC, TLSF_MALLOC, DBG_SR_MEMORY, USE_FUTEX, FAST_LOCK-ADAPTIVE_WAIT, USE_DNS_CACHE, USE_DNS_FAILOVER, USE_NAPTR, USE_DST_BLOCKLIST, HAVE_RESOLV_RES, TLS_PTHREAD_MUTEX_SHARED
ADAPTIVE_WAIT_LOOPS 1024, MAX_RECV_BUFFER_SIZE 262144, MAX_URI_SIZE 1024, BUF_SIZE 65535, DEFAULT PKG_SIZE 8MB
poll method support: poll, epoll_lt, epoll_et, sigio_rt, select.
id: 58499a -dirty
compiled on 11:23:55 May  8 2024 with gcc 9.4.0

```

* **Operating System**:

```
lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.6 LTS
Release:        20.04
Codename:       focal

```

**Comment by deepak-git-account** [comment]
Thanks for the reply.. Both are set to Contact: * and Expires: 0. But Kamailio PCSCF is giving error:
May 13 19:47:22 openims-tvmcfg-node-focal pcscf/pcscf[35816]: ERROR: ims_registrar_pcscf [save.c:318]: save_pending(): No contact headers

**Comment by xkaraman** [comment]
@deepak-git-account 
`save_pending()` does not seem to handle de-registrations according to function code https://github.com/kamailio/kamailio/blob/8363208fff7c101a2779ff64783e5b7dbf8ffd7c/src/modules/ims_registrar_pcscf/save.c#L354. 

Can you maybe use the `save()` function? It needs to be called from onreply_route though. It should handle the `*`.
https://github.com/kamailio/kamailio/blob/8363208fff7c101a2779ff64783e5b7dbf8ffd7c/src/modules/ims_registrar_pcscf/save.c#L566

**Comment by miconda** [comment]
I understand that this is about the ims modules, not about registrar/usrloc modules for the classic SIP.

The IMS modules are not full implementation of the specs, and given that not many are contributing to them, likely this is a missing feature.

One should write the code for it and make a pull request. I changed the title accordingly and if nobody commits to do it soon, this issue will be closed.

**Comment by vingarzan** [comment]
I agree with @deepak-git-account and in general that IETF RFC functionality should be implemented also for IMS domain. And it would be good if the P-CSCF registrar would delete all Contacts locally.

But... things are a bit more complex. 3GPP TS 24.229 does specify a behavior of "*" Contact, yet not for the P-CSCF, only for the S-CSCF. Or??

So IMHO this operation is **not fully applicable at a single P-CSCF**. You can have multiple phones registering the same identity (IMPU). Since not all these phones roam in the same networks, a de-register "*" would anyway only have the proper full effect at the S-CSCF, which would then push `reg-info` notifications to all the P-CSCF instances serving those IMPUs and having Contacts stored in their local reverse-registrars.

**tl;dr - a de-registration with "*" would have a limited local P-CSCF instance effect, hence it's probably better that it's not specified at P-CSCF, to avoid confusion. The S-CSCF-triggered `reginfo` notifications should cleanup all the state in all the P-CSCF instances.**

How could one then trigger a complete de-registration? There are valid use-cases... but there are also administrative de-registration procedures to achieve this triggered from network-side... I'm genuinely curious, since I'm also bothered by the garbage left behind by IMS devices re-attaching. But I'm thinking that the bugs might be at the S-CSCF (e.g. see https://github.com/kamailio/kamailio/pull/3895 for `reginfo` subscribers cleanups) ... 

Do you have a phone which can trigger this?

**Comment by vingarzan** [comment]
Errm... I think we would need a fix first :wink: ...

@miconda hinted that patches would be welcomed. So to spell it out: please contribute.
